### PR TITLE
Allow using OpenSSL in FIPS mode

### PR DESCRIPTION
**Pull request by legoscia**
This is essentially a rebase of #377 by @dszoboszlay on top of 19.1. This change is much smaller, since the changes to make the crypto library use the EVP interface have been merged already.

The aim of this pull request is to allow Erlang/OTP to use OpenSSL in FIPS-140 mode, in order to satisfy specific security requirements (mostly by different parts of the US federal government). While it would be possible to maintain this changeset outside the official source tree, I believe that this would be a valuable addition to Erlang/OTP itself, desired by several different users.

As discussed in the previous pull request, it is true that this feature in a way binds Erlang to OpenSSL (as opposed to e.g. BoringSSL or LibreSSL), but then again, so does FIPS-140 itself: someone who needs this wouldn't be able to use an alternative SSL library anyway. This change is "hidden" behind a configure flag and a test for whether the function `FIPS_mode_set` is provided by the SSL library being used, and should thus have a minimal impact on other users.

### CODE REVIEW

File: erts/configure.in [hunk 1]
```diff
@@ -3905,6 +3905,7 @@ dnl use "PATH/include" and "PATH/lib".
 AC_SUBST(SSL_INCLUDE)
 AC_SUBST(SSL_INCDIR)
 AC_SUBST(SSL_LIBDIR)
+AC_SUBST(SSL_DEFINE)
```
Comment (HansN) [inline erts/configure.in]: @legoscia We would prefer renaming SSL_DEFINE to SSL_FLAGS  to make it similar to other flags.

File: lib/crypto/test/crypto_SUITE.erl [hunk 2]
```diff
@@ -2335,7 +2546,7 @@ ecdh() ->
                  TestCases).
 
 dh() ->
-    {dh, 0087761979513264537414556992123116644042638206717762626089877284926656954974893442000747478454809111207351620687968672207938731607963470779396984752680274820156266685080223616226905101126463253150237669547023934604953898814222890239130021414026118792251620881355456432549881723310342870016961804255746630219, 2}.
+    {dh, 90970053988169282502023478715631717259407236400413906591937635666709823903223997309250405131675572047545403771567755831138144089197560332757755059848492919215391041119286178688014693040542889497092308638580104031455627238700168892909539193174537248629499995652186913900511641708112112482297874449292467498403, 2}.
```
Comment (HansN) [inline lib/crypto/test/crypto_SUITE.erl]: @legoscia just curious, why new dh parameters in the test suite?

Comment (legoscia) [inline lib/crypto/test/crypto_SUITE.erl]: In FIPS mode, there are stricter key length limit. The old key causes this error:

    error:0506706E:Diffie-Hellman routines:GENERATE_KEY:key size too small

**Comment by legoscia** [comment]
Rebased, and changed `SSL_DEFINE` to `SSL_FLAGS`.

File: lib/crypto/test/crypto_SUITE.erl (hunk 2 above)
Comment (HansN) [inline lib/crypto/test/crypto_SUITE.erl]: I see. Quite obvious :)

File: lib/crypto/doc/src/fips.xml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    incompatible with regular builds when compiling crypto with FIPS
+    support.</p>
+  </section>
+
+  <section>
+    <title>Common caveats</title>
+    <p>In FIPS mode non-validated algorithms are disabled. This may
+    cause some unexpected problems in application relying on
+    crypto.</p>
+    <warn><p>Do not try to work around these problems by using
```
Comment (legoscia) [inline lib/crypto/doc/src/fips.xml]: I pushed a fix for this.

**Comment by legoscia** [comment]
Note that even when Erlang is configured with `--enable-fips`, the expected result is still that the FIPS test cases will be skipped, unless the OpenSSL library has been compiled with FIPS support and linked against the OpenSSL FIPS module.

**Comment by HansN** [comment]
I checked the `crypto:info_lib()` response when selecting a machine, which was {`<<"OpenSSL">>,268439711,<<"OpenSSL 1.0.1i-fips 6 Aug 2014">>}`

Do you know if the `-fips` suffix is enough to assure that the two conditions for OpenSSL lib is met?

**Comment by legoscia** [comment]
Yes, the `-fips` suffix seems to be present only when OpenSSL was properly configured and linked for FIPS support.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by HansN** [comment]
Graduated to OTP rel 20.

**Pull request merged**